torch/csrc/jit/script/compiler.cpp

```diff
@@ -1748,15 +1750,32 @@ struct to_ir {
       auto forked = emitSugaredExpr(Expr(trees[0]), 1);
       TreeList sliced_trees(trees.begin() + 1, trees.end());
       auto inputs = getNamedValues(sliced_trees, true);
+      auto attributes = emitAttributes(apply.attributes());
       return emitForkExpr(loc, forked, inputs, attributes);
+    } else if (auto annotate_value = dynamic_cast<AnnotateValue*>(sv.get())) {
+      if (apply.inputs().size() != 2) {
+        throw ErrorReport(loc) << "expected exactly two arguments to attribute but found " << apply.inputs().size();
+      }
+      if (apply.attributes().size() > 0) {
+        throw ErrorReport(loc) << "attribute takes no keyword arguments";
+      }
+      TypePtr type = parseTypeFromExpr(apply.inputs()[0]);
+      Value* expr = emitExpr(apply.inputs()[1], type);
+      if (!expr->type()->isSubtypeOf(type)) {
+        throw ErrorReport(apply.inputs())
+            << "expected an expression of type " << type->python_str()
+            << " but found " << expr->type()->python_str();
+      }
+      return std::make_shared<SimpleValue>(expr);
     } else {
       auto inputs = getNamedValues(apply.inputs(), true);
+      auto attributes = emitAttributes(apply.attributes());
       return sv->call(loc, method, inputs, attributes, n_binders);
     }
   }
 
-  Value* emitExpr(Expr tree) {
-    return emitSugaredExpr(tree, 1)->asValue(tree.range(), method);
+  Value* emitExpr(Expr tree, TypePtr type_hint = nullptr) {
```

Comment:
We're passing `type_hint` deep into the call stack when it's only used in a single place. Can we just emit the ListConstruct expr directly?

I understand that we might want to extend this type annotation to other exprs in the future, but I think having the type hint be inconsistently applied in the meantime is probably net more confusing.